### PR TITLE
Keys API proposal

### PR DESCRIPTION
**Pull request by maririos**

### CODE REVIEW

File: sdk/keyvault/Azure.Security.Keyvault.Keys/src/RSAKey.cs
Comment (schaabs) [inline sdk/keyvault/Azure.Security.Keyvault.Keys/src/RSAKey.cs]: I'm not sure that I wan't to create specific classes for RSAKey and ECKey at this point.  Also it seems odd if we had these classes that they wouldn't be derived from key.

File: sdk/keyvault/Azure.Security.Keyvault.Keys/src/ECKey.cs
Comment (schaabs) [inline sdk/keyvault/Azure.Security.Keyvault.Keys/src/ECKey.cs]: >public class ECKey [](start = 4, length = 18)

See my note on RSA Key

File: sdk/keyvault/Azure.Security.Keyvault.Keys/src/RSAKey.cs
Comment (schaabs) [inline sdk/keyvault/Azure.Security.Keyvault.Keys/src/RSAKey.cs]: >public JsonWebKey KeyMaterial { get; set; } [](start = 8, length = 43)

If this class is intended as an input bag for CreateKey it shouldn't have KeyMaterial on it.  Only a handful of properties on the JWK are actually settable at creation time.

File: sdk/keyvault/Azure.Security.Keyvault.Keys/src/KeyClient.cs
Comment (schaabs) [inline sdk/keyvault/Azure.Security.Keyvault.Keys/src/KeyClient.cs]: should be GetAllDeletedKeysAsync

File: sdk/keyvault/Azure.Security.Keyvault.Keys/src/KeyClient.cs
Comment (schaabs) [inline sdk/keyvault/Azure.Security.Keyvault.Keys/src/KeyClient.cs]: >GetBatchDeletedKeys [](start = 57, length = 19)

For secrets we used the verbiage 'all' rather than 'batch', so this should be GetAllDeletedKeys

File: sdk/keyvault/Azure.Security.Keyvault.Keys/src/KeyClient.cs
Comment (schaabs) [inline sdk/keyvault/Azure.Security.Keyvault.Keys/src/KeyClient.cs]: Need to add a few missing APIs:
  - GetAll / GetAllAsync
  - GetAllVerions / GetAllVerionsAsync

File: sdk/keyvault/Azure.Security.Keyvault.Keys/src/KeyBase.cs
Comment (schaabs) [inline sdk/keyvault/Azure.Security.Keyvault.Keys/src/KeyBase.cs]: >public string Version { get; set; } [](start = 8, length = 35)

Not settable

File: sdk/keyvault/Azure.Security.Keyvault.Keys/src/KeyBase.cs
Comment (schaabs) [inline sdk/keyvault/Azure.Security.Keyvault.Keys/src/KeyBase.cs]: >public string Name { get; set; } [](start = 8, length = 32)

Not settable

File: sdk/keyvault/Azure.Security.Keyvault.Keys/src/KeyBase.cs
Comment (schaabs) [inline sdk/keyvault/Azure.Security.Keyvault.Keys/src/KeyBase.cs]: >public string Id { get; set; } [](start = 8, length = 30)

Not settable after construction

File: sdk/keyvault/Azure.Security.Keyvault.Keys/src/KeyBase.cs
Comment (schaabs) [inline sdk/keyvault/Azure.Security.Keyvault.Keys/src/KeyBase.cs]: >public string VaultId { get; set; } [](start = 8, length = 35)

not sure what this is intended to be?

File: sdk/keyvault/Azure.Security.Keyvault.Keys/src/KeyBase.cs
Comment (schaabs) [inline sdk/keyvault/Azure.Security.Keyvault.Keys/src/KeyBase.cs]: >public string VaultUrl { get; set; } [](start = 8, length = 36)

Should be a Uri rather than string, and not settable after construction

File: sdk/keyvault/Azure.Security.Keyvault.Keys/src/Key.cs
Comment (schaabs) [inline sdk/keyvault/Azure.Security.Keyvault.Keys/src/Key.cs]: Need to add constructors

File: sdk/keyvault/Azure.Security.Keyvault.Keys/src/JsonWebKey.cs
Comment (pakrym) [inline sdk/keyvault/Azure.Security.Keyvault.Keys/src/JsonWebKey.cs]: I always wondered why does it have Json and Web in the name?

File: sdk/keyvault/Azure.Security.Keyvault.Keys/src/DeletedKey.cs
Comment (pakrym) [inline sdk/keyvault/Azure.Security.Keyvault.Keys/src/DeletedKey.cs]: Is this the right `Keyvault` casing? Same for the project name.

File: sdk/keyvault/Azure.Security.Keyvault.Keys/src/Azure.Security.Keyvault.Keys.csproj
Comment (pakrym) [inline sdk/keyvault/Azure.Security.Keyvault.Keys/src/Azure.Security.Keyvault.Keys.csproj]: This is an old package, we've moved to using `<PackageReference Include="System.Text.Json" Version="4.6.0-preview6.19226.8" />`

File: sdk/keyvault/Azure.Security.Keyvault.Keys/src/Azure.Security.Keyvault.Keys.csproj
Comment (pakrym) [inline sdk/keyvault/Azure.Security.Keyvault.Keys/src/Azure.Security.Keyvault.Keys.csproj]: Use the props to import Azure.Core

```
  <Import Project="$(MSBuildThisFileDirectory)..\..\..\core\Azure.Core\src\Azure.Core.props" />
```

File: sdk/keyvault/Azure.Security.Keyvault.Keys/src/Azure.Security.Keyvault.Keys.csproj
Comment (pakrym) [inline sdk/keyvault/Azure.Security.Keyvault.Keys/src/Azure.Security.Keyvault.Keys.csproj]: @weshaggard can we make this false by default for track 2 SDKs?

File: sdk/keyvault/Azure.Security.Keyvault.Keys/src/KeyBase.cs
Comment (pakrym) [inline sdk/keyvault/Azure.Security.Keyvault.Keys/src/KeyBase.cs]: nit: formatting

File: sdk/keyvault/Azure.Security.Keyvault.Keys/src/Azure.Security.Keyvault.Keys.csproj
Comment (weshaggard) [reply sdk/keyvault/Azure.Security.Keyvault.Keys/src/Azure.Security.Keyvault.Keys.csproj]: Yes or we can update the default reference set to make sense for track 2 libraries.

Comment (weshaggard) [reply sdk/keyvault/Azure.Security.Keyvault.Keys/src/Azure.Security.Keyvault.Keys.csproj]: Don't worry about it as part of this PR. I'll clean these up as well as some of the other properties I want to centralize.

Comment (maririos) [reply sdk/keyvault/Azure.Security.Keyvault.Keys/src/Azure.Security.Keyvault.Keys.csproj]: Once your PR is merged I can clean up this project to follow the other ones

Comment (weshaggard) [reply sdk/keyvault/Azure.Security.Keyvault.Keys/src/Azure.Security.Keyvault.Keys.csproj]: I just merged it. Feel free to clean-up this project similar to what I did in the other projects.

Comment (maririos) [reply sdk/keyvault/Azure.Security.Keyvault.Keys/src/Azure.Security.Keyvault.Keys.csproj]: Will do

File: sdk/keyvault/Azure.Security.Keyvault.Keys/src/Azure.Security.Keyvault.Keys.csproj
Comment (maririos) [reply sdk/keyvault/Azure.Security.Keyvault.Keys/src/Azure.Security.Keyvault.Keys.csproj]:
```
Warning	NU1603	[...] System.Text.Json 4.6.0-preview6.19226.8 was not found. An approximate best match of System.Text.Json 4.6.0-preview6.19259.10 was resolved.
```
I updated the reference to `System.Text.Json 4.6.0-preview6.19259.10`

File: sdk/keyvault/Azure.Security.Keyvault.Keys/src/DeletedKey.cs
Comment (maririos) [reply sdk/keyvault/Azure.Security.Keyvault.Keys/src/DeletedKey.cs]: Ohh thanks.

File: sdk/keyvault/Azure.Security.Keyvault.Keys/src/JsonWebKey.cs
Comment (schaabs) [reply sdk/keyvault/Azure.Security.Keyvault.Keys/src/JsonWebKey.cs]: It's just what it's called in the JOSE standard.  JWK (json web key), JWE (json web encryption), JWS (json web signature), JWT (json web token).  https://tools.ietf.org/html/rfc7517

File: .gitignore
Comment (AlexGhiondea) [inline .gitignore]: KeyVault

File: sdk/keyvault/Azure.Security.Keyvault.Keys/src/KeyClient.cs
Comment (maririos) [reply sdk/keyvault/Azure.Security.Keyvault.Keys/src/KeyClient.cs]: For configuration service there was a discussion about it, and the conclusion was to use batch instead of all.
I'm going to renaming it and delete the batch and all

File: sdk/keyvault/Azure.Security.Keyvault.Keys/src/RSAKey.cs
Comment (maririos) [reply sdk/keyvault/Azure.Security.Keyvault.Keys/src/RSAKey.cs]: changed after our conversation on Teams about having `KeyCreateOptions`

File: sdk/keyvault/Azure.Security.Keyvault.Keys/src/KeyClient.cs
Comment (maririos) [reply sdk/keyvault/Azure.Security.Keyvault.Keys/src/KeyClient.cs]: added

File: sdk/keyvault/Azure.Security.Keyvault.Keys/src/KeyBase.cs
Comment (maririos) [reply sdk/keyvault/Azure.Security.Keyvault.Keys/src/KeyBase.cs]: yeah, my bad, was supposed to be Uri

File: sdk/keyvault/Azure.Security.Keyvault.Keys/src/KeyClient.cs
Comment (schaabs) [reply sdk/keyvault/Azure.Security.Keyvault.Keys/src/KeyClient.cs]: Also missing ImportKey, ImportKeyAsync

---
In reply to: [287178615](https://github.com/Azure/azure-sdk-for-net/pull/6374#discussion_r287178615) [](ancestors = 287178615)

File: sdk/keyvault/Azure.Security.Keyvault.Keys/src/KeyClient.cs
Comment (schaabs) [inline sdk/keyvault/Azure.Security.Keyvault.Keys/src/KeyClient.cs]: >GetKeysAsync [](start = 55, length = 12)

Should be GetAllKeysAsync

Comment (schaabs) [inline sdk/keyvault/Azure.Security.Keyvault.Keys/src/KeyClient.cs]: >int maxresults = 25 [](start = 68, length = 19)

Remove this parameter

File: sdk/keyvault/Azure.Security.Keyvault.Keys/src/KeyClient.cs
Comment (schaabs) [inline sdk/keyvault/Azure.Security.Keyvault.Keys/src/KeyClient.cs]: >int maxresults = 25 [](start = 58, length = 19)

Remove this parameter

Comment (schaabs) [inline sdk/keyvault/Azure.Security.Keyvault.Keys/src/KeyClient.cs]: >GetKeys [](start = 50, length = 7)

Should be GetAllKeys

File: sdk/keyvault/Azure.Security.Keyvault.Keys/src/KeyClient.cs
Comment (schaabs) [inline sdk/keyvault/Azure.Security.Keyvault.Keys/src/KeyClient.cs]: >int maxresults = 25 [](start = 78, length = 19)

Remove this parameter

File: sdk/keyvault/Azure.Security.Keyvault.Keys/src/KeyClient.cs
Comment (schaabs) [inline sdk/keyvault/Azure.Security.Keyvault.Keys/src/KeyClient.cs]: >int maxresults = 25 [](start = 88, length = 19)

Remove this parameter

File: sdk/keyvault/Azure.Security.Keyvault.Keys/src/KeyClient.cs
Comment (schaabs) [inline sdk/keyvault/Azure.Security.Keyvault.Keys/src/KeyClient.cs]: >int maxresults = 25 [](start = 82, length = 19)

Remove maxresults parameter

File: sdk/keyvault/Azure.Security.Keyvault.Keys/src/KeyClient.cs
Comment (schaabs) [inline sdk/keyvault/Azure.Security.Keyvault.Keys/src/KeyClient.cs]: >int maxresults = 25 [](start = 72, length = 19)

remove maxresults parameter

File: sdk/keyvault/Azure.Security.Keyvault.Keys/src/KeyClient.cs
Comment (schaabs) [reply sdk/keyvault/Azure.Security.Keyvault.Keys/src/KeyClient.cs]: In the case of the configuration service there is filtering, so you're not necessarily returning all, so batch makes sense.  This is not the case for key vault all entities will be returned always, hence all. 

---
In reply to: [287145821](https://github.com/Azure/azure-sdk-for-net/pull/6374#discussion_r287145821) [](ancestors = 287145821)

File: sdk/keyvault/Azure.Security.Keyvault.Keys/src/KeyClient.cs
Comment (schaabs) [inline sdk/keyvault/Azure.Security.Keyvault.Keys/src/KeyClient.cs]: >string name [](start = 48, length = 11)

this param should be byte[] keyBackup

File: sdk/keyvault/Azure.Security.Keyvault.Keys/src/KeyClient.cs
Comment (schaabs) [inline sdk/keyvault/Azure.Security.Keyvault.Keys/src/KeyClient.cs]: >string name [](start = 65, length = 11)

this param should be byte[] keyBackup

File: sdk/keyvault/Azure.Security.Keyvault.Keys/src/KeyBase.cs
Comment (schaabs) [inline sdk/keyvault/Azure.Security.Keyvault.Keys/src/KeyBase.cs]: >VaultUrl [](start = 19, length = 8)

should be called VaultUri

**Comment by maririos** [comment]
@schaabs  PR updated with latest feedback

**Comment by maririos** [comment]
@schaabs could I get this PR approved to I can merge it?
I will address tasks from the API Review meeting in future PRs

**Review by schaabs — APPROVED** [review]

**Comment by schaabs** [comment]
approving now,  Changes stemming from API review board comments will be addressed in separate issues / PRs

**Pull request merged**
